src/components/preferences/Preferences.js

```diff
@@ -0,0 +1,466 @@
+            onError: (e) => {
+                showErrorAnnouncer(formatMessage(intl, "redirectError"), e);
+            },
+        },
+    });
+
+    const { isFetching: isFetchingStat } = useQuery({
+        queryKey: fetchDetailsKey,
+        queryFn: getResourceDetails,
+        config: {
+            onSuccess: (resp) => {
+                const details = resp?.paths[defaultOutputFolder];
+                if (!isWritable(details?.permission)) {
+                    setOutputFolderValidationError(
+                        formatMessage(intl, "permissionSelectErrorMessage")
+                    );
+                } else {
+                    setOutputFolderValidationError(null);
+                }
+            },
+        },
+        onError: (e) => {
+            setOutputFolderValidationError(
+                formatMessage(intl, "permissionSelectErrorMessage")
+            );
+        },
+    });
+
+    const handleSubmit = (values) => {
+        //preven dupe submission
+        if (prefMutationStatus !== constants.LOADING) {
+            if (outputFolderValidationError) {
+                announce({
+                    text: formatMessage(intl, "validationMessage"),
+                });
+            } else {
+                const updatedPref = values;
+                updatedPref.default_output_folder.id =
+                    values.defaultOutputFolder;
+                updatedPref.default_output_folder.path =
+                    values.defaultOutputFolder;
+                preferences(updatedPref);
+            }
+        }
+    };
+
+    const restoreDefaults = (setFieldValue) => (event) => {
+        setFieldValue(prefConstants.keys.REMEMBER_LAST_PATH, true);
+        setFieldValue(
+            prefConstants.keys.ENABLE_ANALYSIS_EMAIL_NOTIFICATION,
+            true
+        );
+        setFieldValue(prefConstants.keys.ENABLE_WAIT_TIME_MESSAGE, true);
+        setFieldValue(
+            prefConstants.keys.ENABLE_IMPORT_EMAIL_NOTIFICATION,
+            true
+        );
+        setFieldValue(prefConstants.keys.ENABLE_HPC_PROMPT, true);
+        setFieldValue(
+            prefConstants.keys.NOTIFICATION_KB_SC,
+            prefConstants.defaults.NOTIFICATION_SC
+        );
+        setFieldValue(
+            prefConstants.keys.DATA_KB_SC,
+            prefConstants.defaults.DATA_SC
+        );
+        setFieldValue(
+            prefConstants.keys.APPS_KB_SC,
+            prefConstants.defaults.APPS_SC
+        );
+        setFieldValue(
+            prefConstants.keys.ANALYSES_KB_SC,
+            prefConstants.defaults.ANALYSES_SC
+        );
+        setFieldValue(
+            prefConstants.keys.DEFAULT_OUTPUT_FOLDER,
+            userPref.system_default_output_dir.path
+        );
+        setShowRestoreConfirmation(false);
+    };
+
+    const setDefaultFolder = (setFieldValue, newFolder) => {
+        setFieldValue(prefConstants.keys.DEFAULT_OUTPUT_FOLDER, newFolder);
+        setDefaultOutputFolder(newFolder);
+    };
```

Comment:
I tried looking at the `react-query` docs to see if a query can be used as a promise, since that's the only way to do async validation with Formik, but I didn't find a way.

I'm not sure if you tried something like this, but maybe one way to let Formik handle validation, instead of manually setting it with state, is call `setFetchDetailsKey` here, instead of calling `setFieldValue` right away.

Then in the  `onSuccess` callback of that query, the entire folder details object can be set with `setFieldValue`. Once that is called it will trigger validation, and then inside `validateShortCuts` (which should be renamed) we can check for write permissions and set an error message if necessary.

This would also require using `defaultOutputFolder.path` for the `value` of the text field for display in the form, and also in `handleSubmit`.

Reply:
Good Idea. Let me give this a shot!

Reply:
Welp. Doesn't seem to work. I also see roadblock with this approach. For example, when I try to restore defaults, that causes exceptions.
`A component is changing an uncontrolled input of type text to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa)` 

May be we can revisit this when we add `webhooks` preferences.

Reply:
When you got the "uncontrolled input" warning, were you still defaulting the `value` of the text field to an empty string if there is no folder path?
`value={defaultOutputFolder?.path || ""}`

Reply:
Yes.